### PR TITLE
fix(background/TabState): use `tabId` as `tab` isn't a live reference to tab

### LINKED ISSUE #286: [Background] Overpaying
Ordering: this ticket should ideally follow #145 

If the receiver’s wallet cannot receive AMOUNT, we should send the minimum amount that can be sent. If the receiver’s asset scale is 2, we will not be able to send SPLITTED_AMOUNT, and we need to send the smallest unit 1 + 1 (we can not send 1 - cause: @interledger/pay library, used in Rafiki). We call this action OVERPAYING. We also need to adjust the payment interval. By default, the happy path, we send a payment every second. If overpaying, we need to adjust the interval as well: We need to send a payment every 30 seconds (dynamic - it will not always be 30 seconds, we need to calculate the interval based on the smallest amount that can be sent) to facilitate RATE_OF_PAY. 

- set the amount as minimum in the constructor or when recalculating the rate.
- if overpaying, calculate the interval as well:  0.02 * 3600/0.60 = 120s (2 minutes)

When a payment session is overpaying we need to store the last payment timestamp. Reasoning: if a user keeps refreshing a page and we have a payment session that is overpaying, we will send $0.02 whenever a payment session is established for the wallet address . If we have the timestamp, we can check if the page has received a payment before and if CURRENT_TIMSTAMP is not greater than the payment timestamp we will wait for CURRENT_TIMESTAMP - PAYMENT_TIMESTAMP until sending the next payment.

Whenever a payment session is overpaying save:
- Website url (we store the sha1 the URL)
- Tab ID
- Wallet address
- Last payment timestamp
- Expires At = Last payment timestamp + MS (TBD)

Whenever we receive the START_MONETIZATION event, we check the overpaying state for the current tab + website url. If a record is found, initialize a new payment session but do not start it. We need to wait for LAST_PAYMENT_TIMESTAMP + INTERVAL (dynamic for each payment session that is overpaying) - CURRENT_TIMESTAMP

### LINKED ISSUE #287: [Background] [TO REFINE] Clear outdated data regarding overpaying
Every X seconds, run a function that checks the expires at property for an overpaying session. If CURRENT_TIMESTAMP > EXPIRES_AT delete the session from state.

Related to #286

### PR DESCRIPTION
**Pull request by dianafulga**
Closes #286
Closes #287 

The data is deleted immediately after the tab closes because reopening the tab (cmd + shift + T) will generate a new tabId anyway.

### CODE REVIEW

**Comment by raducristianpopa** [comment]
<h2>Extension builds preview</h2>

<table role="table">
<thead>
  <tr>
    <th>Name</th>
    <th>Link</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td align="center">Latest commit</td>
    <td align="center">4361c2b2a3783d1b23ab665c9d01278613bc1c26</td>
  </tr>
  <tr>
    <td align="center">Latest job logs</td>
    <td align="center"><a href="https://github.com/interledger/web-monetization-extension/actions/runs/10005285870">Run #10005285870</a></td>
  </tr>
  <tr><td align="center">Badge</td><td align="center"><a href="https://github.com/interledger/web-monetization-extension/suites/26192175239/artifacts/1718327493">Download</a></td></tr><tr><td align="center">Badge</td><td align="center"><a href="https://github.com/interledger/web-monetization-extension/suites/26192175239/artifacts/1718327495">Download</a></td></tr>
</tbody>
</table>

**Comment by sidvishnoi** [comment]
Can you clarify/confirm two things:
1. Refreshing certainly gives same tabId across browsers? I mean Firefox and Chrome only for now.
1. Does this also work cross browser? 
   > reopening the tab (cmd + shift + T) will generate a new tabId anyway

File: src/background/services/monetization.ts [hunk 1]
```diff
@@ -222,6 +221,9 @@ export class MonetizationService {
     }
 
     delete this.sessions[tabId]
+
+    this.tabState.clearByTabId(tabId)
+
```
Comment (sidvishnoi) [inline src/background/services/monetization.ts:226]: Nit:
```suggestion
    this.tabState.clearByTabId(tabId)
```

File: src/background/services/tabState.ts [hunk 2]
```diff
@@ -14,7 +13,7 @@ interface SaveOverpayingDetails {
 }
 
 export class TabState {
-  private state = new WeakMap<Tabs.Tab, Map<string, State>>()
+  private state = new Map<number, Map<string, State>>()
```
Comment (sidvishnoi) [inline src/background/services/tabState.ts:16]: So we're clear what the number represents.
Also consider providing same for `string` (as `SessionId`, but not relevant to this PR)

```suggestion
  type TabId = NonNullable<Tabs.Tab['id']>
  // ...
  private state = new Map<TabId, Map<string, State>>()
```

**Review by sidvishnoi — COMMENTED** [review]
LGTM. Some nits and that cross-browser concern to be addressed.

**Comment by sidvishnoi** [comment]
> > reopening the tab (cmd + shift + T) will generate a new tabId anyway
> 
> Does this work cross browser? 
> 
> 
>     

@dianafulga does it work same in Firefox?

**Comment by sidvishnoi** [comment]
I checked and can confirm this works in Firefox also.

File: src/background/services/tabState.ts [hunk 3]
```diff
@@ -1,5 +1,5 @@
 import type { MonetizationEventDetails } from '@/shared/messages'
-import type { Tabs } from 'webextension-polyfill'
+import { Tabs } from 'webextension-polyfill'
```
Comment (sidvishnoi) [inline src/background/services/tabState.ts:2]:
```suggestion
import type { Tabs } from 'webextension-polyfill'
```

File: src/background/services/tabState.ts [hunk 4]
```diff
@@ -23,12 +26,12 @@ export class TabState {
   }
 
   getOverpayingDetails(
-    tab: Tabs.Tab,
+    tabId: number,
```
Comment (sidvishnoi) [inline src/background/services/tabState.ts:29]:
```suggestion
    tabId: TabId,
```

File: src/background/services/tabState.ts [hunk 5]
```diff
@@ -44,7 +47,7 @@ export class TabState {
   }
 
   saveOverpaying(
-    tab: Tabs.Tab,
+    tabId: number,
```
Comment (sidvishnoi) [inline src/background/services/tabState.ts:50]:
```suggestion
    tabId: TabId,
```

File: src/background/services/tabState.ts [hunk 6]
```diff
@@ -55,19 +58,23 @@ export class TabState {
...
       })
-      this.state.set(tab, tabState)
+      this.state.set(tabId, tabState)
     } else {
       state.expiresAtTimestamp = expiresAtTimestamp
       state.lastPaymentTimestamp = now
     }
   }
+
+  clearByTabId(tabId: number) {
```
Comment (sidvishnoi) [inline src/background/services/tabState.ts:77]:
```suggestion
  clearByTabId(tabId: TabId) {
```

File: src/background/services/tabState.ts [hunk 7]
```diff
@@ -13,8 +13,11 @@ interface SaveOverpayingDetails {
   intervalInMs: number
 }
 
+type TabId = NonNullable<Tabs.Tab['id']>
+type SessionId = string
+
 export class TabState {
-  private state = new WeakMap<Tabs.Tab, Map<string, State>>()
+  private state = new Map<TabId, Map<SessionId, State>>()
```
Comment (sidvishnoi) [inline src/background/services/tabState.ts:20]:
```suggestion
  private state = new Map<TabId, Map<string, State>>()
```

File: src/background/services/tabState.ts [hunk 8]
```diff
@@ -13,8 +13,11 @@ interface SaveOverpayingDetails {
   intervalInMs: number
 }
 
+type TabId = NonNullable<Tabs.Tab['id']>
+type SessionId = string
```
Comment (sidvishnoi) [inline src/background/services/tabState.ts:17]:
```suggestion
```

**Review by sidvishnoi — APPROVED** [review]

**Pull request merged**